Source/DafnyLanguageServer.Test/Various/PluginsTestBase.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.Dafny.LanguageServer.IntegrationTest.Util;
+using Microsoft.Extensions.Configuration;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+
+namespace Microsoft.Dafny.LanguageServer.IntegrationTest.Various;
+
+public abstract class PluginsTestBase : DafnyLanguageServerTestBase {
+  protected ILanguageClient Client;
+  protected DiagnosticsReceiver DiagnosticReceiver;
+  protected string LibraryPath;
+
+  private static readonly object PluginSync = new { };
+
+  /// <summary>
+  /// This method creates a library and returns the path to that library.
+  /// The library extends a Rewriter so that we can verify that Dafny invokes it if provided in argument.
+  /// </summary>
+  private string GetLibrary(string code, string assemblyName) {
+    var temp = Path.GetTempFileName();
+    var compilation = CSharpCompilation.Create(assemblyName);
+    var standardLibraries = new List<string>()
+    {
+      "DafnyPipeline",
+      "System.Runtime",
+      "Boogie.Core",
+      "System.Collections"
+    };
+    compilation = compilation.AddReferences(standardLibraries.Select(fileName =>
+        MetadataReference.CreateFromFile(Assembly.Load((string)fileName).Location)))
+      .AddReferences(
+        MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location))
+      .WithOptions(
+        new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
+      );
+    var syntaxTree = CSharpSyntaxTree.ParseText(code);
+    compilation = compilation.AddSyntaxTrees(syntaxTree);
+    var assemblyPath = $"{temp}.dll";
+    var result = compilation.Emit(assemblyPath);
+
+    Assert.IsTrue(result.Success, string.Join("\n", result.Diagnostics.Select(d => d.ToString())));
+    return assemblyPath;
+  }
+
+  protected abstract string GetLibraryCode();
+
+  protected abstract string GetLibraryName();
+
+  protected abstract string[] GetCommandLineArgument();
+
+  public async Task SetUpPlugin() {
+    DiagnosticReceiver = new();
+    LibraryPath = GetLibrary(GetLibraryCode(), GetLibraryName());
+    Monitor.Enter(PluginSync);
```

Comment:
What's the reason for this monitor synchronization? Tests (should) run synchronously by default (other tests would fail miserably if they didn't). Moreover, I believe a `SemaphoreSlim` would be more suitable here.

Reply:
What makes you believe tests should run synchronously by default? I routinely run 8 tests in parallel and the CLI does this too I believe.
The Monitor solves a problem that happens both in the CLI and locally: PluginAdvancedTest and PluginTest both dynamically load a plugin, and DafnyLanguageServer loads plugins only the first time they are provided, so running the tests independently worked, but running them at the same time would prevent the second test to install its plugin.

I don't think `SemaphoreSlim` would be appropriate here because really we only want at most one plugin test to be run at a time during tests, not many. I found this doc here:
https://www.onlinebuff.com/article_understand-monitor-vs-mutex-vs-semaphore-vs-semaphoreslim-onlinebuff_60.html

Reply:
I actually removed the synchronization code since I'm now removing the plugin at the end, and it works now.

Reply:
> What makes you believe tests should run synchronously by default? I routinely run 8 tests in parallel and the CLI does this too I believe.

I thought it was opt-in. The reason I think that tests would fail, if they'd run indeed in parallel, is that the access to Dafny's compiler core is not mutually exclusive. However there's shared memory because of the static fields; thus we have low level data races and race conditions.

Even though it's now obsolete, I still think it's valuable to answer your question:
The reason why I suggested to not use a monitor lock is that it's rather uncommon to split the acquisition and the release of the lock across multiple methods; that's why the `lock` statement is handy here and should be preferred.
You are right that `SemaphoreSlim` allows multiple acquisitions of permits, and depending on the implementation, effectively allowing multiple threads to enter a critical section. However, this is not a problem if there's only a single permit, making the semaphore a binary semaphore that can be used the same as a dedicated mutex type. Moreover, `SemaphoreSlim` provides asynchronous APIs.